### PR TITLE
Project Auth Noemi and Estefanía

### PR DESCRIPTION
**Pull request by esteficodes**
Link to my deployed project: 

Frontend: https://project-auth-week20.netlify.app/

Backend: https://project-auth-week20.herokuapp.com/

My reflections on how this week's project turned out:

We had a great week and pair-programming was wonderful. Everything was working fine in the local version of the project but we are having some issues after deployment.

Things I'd like to have clarified or explained in more detail:

We would like some help regarding the backend not working after deployment. We couldn't figure it out yet.

### CODE REVIEW

File: frontend/src/components/Main.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                    })
+                } else {
+                    dispatch(thoughts.actions.setErrors(data))
+                }
+            })
+    }, [accessToken, dispatch])
+
+    return (
+        <div className="main-wrapper">
+            <iframe className="happy-thoughts-site" width="100%" height="100%" src="https://happy-thoughts-app-estefania.netlify.app/" title="Happy Thoughts site"></iframe>
```
Comment (Irina-web-dev) [inline frontend/src/components/Main.js:54]: This container for the Happy Thoughts website looks too small, maybe try to make it larger ✨🐱‍🏍

image

File: frontend/src/components/Welcome.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        <div>
+            <div className="main-wrapper">
+                <iframe title="Welcome-gif" src="https://giphy.com/embed/l0MYC0LajbaPoEADu" width="350" height="240" frameBorder="0" class="giphy-embed" allowFullScreen></iframe>
+                <p className="text">Congratulation to becoming a member of the coolest crowd there is!</p>
+                <p className="text">Don't get confused by the missing content on here. We are cool silently and inactively.</p>
+                <h3 className="welcome-title">Now login to enter the memberarea <Link to="/login">here</Link></h3>
+            </div>
+        </div>
+    )
+}
```
Comment (Irina-web-dev) [inline frontend/src/components/Welcome.js:53]: Wow, this is really nice, we could see that you had much fun creating this 🥳 Great job, Foxes! 🦊👌
image

File: README.md [hunk 3]
```diff
@@ -1,13 +1,25 @@
...
-Describe how you approached to problem, and what tools and techniques you used to solve it. How did you plan? What technologies did you use? If you had more time, what would be next?
+We learned how to authenticate users using tokens and how to store passwords in databases in a secure way. We also created a Frontend with a form to register, sign in and be able to see restricted content in case the login was successful.
+
+We used MongodB and Mongoose to build our Backend Express API. This API has 3 endpoints: 
+
+<ul>
+  <li>A POST endpoint to create a new user</li>
+  <li>A POST endpoint for the existing users to log in</li>
+  <li>A GET restricted endpoint which is the home of the restricted site's content. It uses an authentication middleware, so it's only accessible with a valid access token.</li>
+</ul>
```
Comment (Irina-web-dev) [inline README.md:17]: Thank you for this summary. It makes it super clear what to look for in your code 👍🔅

**Review by Irina-web-dev — COMMENTED** [review]
Hi Noemi and Estefania! 🦊🦊

We enjoyed doing your code review 🔅 Your full stack project looks great and works like a clock 🕑  We tested everything and we have only one small suggestion for improvement, you can find it down in the code.

Your Foxy Ladies Irina & Daniela 🦊🦊

File: README.md [hunk 4]
```diff
@@ -1,13 +1,25 @@
...
+
+The Frontend was built using a combination of React/Redux and local states in order to control the login/signup flow. 
 
 ## View it live
 
-Every project should be deployed somewhere. Be sure to include the link to the deployed project so that the viewer can click around and see what it's all about.
+Frontend: https://project-auth-week20.netlify.app/
```
Comment (KarolCislowski) [inline README.md:23]: To fix it you need to create `_redirects` file in the `public` directory of your react project and paste it inside:
``` 
/* /index.html 200
```

image

File: backend/server.js [hunk 5]
```diff
@@ -1,26 +1,112 @@
...
+    unique: true
+  },
+  password: {
+    type: String,
+    required: true
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString('hex')
+  }
```
Comment (KarolCislowski) [inline backend/server.js:30]: Nice model 👍🏻

File: frontend/src/components/Login.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        }
+        fetch(API_URL(mode), options)
+          .then(res => res.json())
+          .then(data => {
+             if (data.success) {
+                batch(() => {//batch updates all the dispatches at the same time, not one at a time
+                    dispatch(user.actions.setUsername(data.username))
+                    dispatch(user.actions.setAccessToken(data.accessToken))
+                    dispatch(user.actions.setErrors(null))
+                })
```
Comment (KarolCislowski) [inline frontend/src/components/Login.js:43]: nice batch() 👍🏻

File: frontend/src/components/Main.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                if (data.success) {
+                    batch(() => {
+                        dispatch(thoughts.actions.setThoughts(data.thoughts))
+                        dispatch(thoughts.actions.setErrors(null))
+                    })
+                } else {
+                    dispatch(thoughts.actions.setErrors(data))
+                }
+            })
+    }, [accessToken, dispatch])
```
Comment (KarolCislowski) [inline frontend/src/components/Main.js:50]: why 2 effects when you can have just one: 
```jsx
useEffect(() => {
    if (!accessToken) {
      history.push('/login')
    } else {
      const options = { ...}
      ...
    }
  }, [accessToken, history, dispatch])
```

File: frontend/src/components/Main.js (hunk 1 above)
Comment (KarolCislowski) [reply frontend/src/components/Main.js:54]: Maybe just use the HT project and add a header with a logout button to it instead container?

File: frontend/src/components/Welcome.js (hunk 2 above)
Comment (KarolCislowski) [reply frontend/src/components/Welcome.js:53]: Love it 😻

File: frontend/src/reducers/user.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        setAccessToken: (store, action) => {
+            store.accessToken = action.payload
+        },
+        setErrors: (store, action) => {
+            store.errors = action.payload
+        }
+    } 
+})
+
+export default user
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:23]: You need to store your data in the local storage to avoid logging out your user after refreshing the page.

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
It is really nice app but you need to persist your user in the local storage to avoid refreshing.


Because your graduation is so soon please let me know on slack (a short note with a link to this PR) when you finish fixing your app.

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
Your code looks nice, but your netlify is not updated for some reason:  
image

**Comment by esteficodes** [comment]
> Your code looks nice, but your netlify is not updated for some reason:
> image

HI @KarolCislowski ! Thanks for the heads up. It should be working fine now. 😃

**Review by KarolCislowski — APPROVED** [review]
Great APP

**Pull request closed without merging**